File: curriculum/11_load_balancer/node-impl/src/__tests__/load-balancer.test.ts
import http, { Server } from 'node:http';
import { once } from 'node:events';
import { afterEach, describe, expect, it } from 'vitest';
import { LoadBalancer, defaultConfig } from '../load-balancer';

const servers: Server[] = [];

async function backend(name: string, status = 200): Promise<{ url: string; hits: string[]; close: () => Promise<void> }> {
  const hits: string[] = [];
  const server = http.createServer((req, res) => {
    if (req.url === '/health') { res.writeHead(status); res.end('ok'); return; }
    hits.push(`${req.method} ${req.url} ${req.headers['x-request-id'] ?? ''}`);
    res.setHeader('x-backend', name);
    res.end(name);
  });
  await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
  servers.push(server);
  const address = server.address();
  if (!address || typeof address === 'string') throw new Error('missing port');
  return { url: `http://127.0.0.1:${address.port}`, hits, close: () => new Promise((resolve) => server.close(() => resolve())) };
}

afterEach(async () => { await Promise.all(servers.splice(0).map((server) => new Promise<void>((resolve) => server.close(() => resolve())))); });

describe('Node load balancer', () => {
  it('round-robins only eligible healthy backends and applies weights', async () => {
    const lb = new LoadBalancer({ ...defaultConfig([{ id: 'a', url: 'http://a', weight: 2 }, { id: 'b', url: 'http://b', weight: 1 }]) });
    lb.markBackendHealthy('a'); lb.markBackendHealthy('b');
    expect([lb.selectBackend()?.id, lb.selectBackend()?.id, lb.selectBackend()?.id]).toEqual(['a', 'a', 'b']);
  });

  it('chooses least active connections with deterministic tie-breaking', () => {
    const lb = new LoadBalancer({ ...defaultConfig([{ id: 'b', url: 'http://b' }, { id: 'a', url: 'http://a' }]), routingAlgorithm: 'least_connections' });
    lb.markBackendHealthy('a'); lb.markBackendHealthy('b');
    expect(lb.selectBackend()?.id).toBe('a');
  });

  it('updates health and circuit breaker state from active and passive failures', async () => {
    const failing = await backend('bad', 500);
    const lb = new LoadBalancer({ ...defaultConfig([{ id: 'bad', url: failing.url }]), failureThreshold: 1 });
    await lb.checkBackend('bad');
    expect(lb.snapshots()[0]).toMatchObject({ health: 'unhealthy', circuitState: 'open' });
  });

  it('proxies method, path, query, headers, and exposes admin metrics', async () => {
    const a = await backend('a');
    const b = await backend('b');
    const lb = new LoadBalancer(defaultConfig([{ id: 'a', url: a.url }, { id: 'b', url: b.url }]));
    lb.markBackendHealthy('a'); lb.markBackendHealthy('b');
    const server = lb.listen(0);
    await once(server, 'listening');
    servers.push(server);
    const address = server.address();
    if (!address || typeof address === 'string') throw new Error('missing port');
    const base = `http://127.0.0.1:${address.port}`;
    const first = await fetch(`${base}/demo?x=1`, { method: 'POST', headers: { 'x-request-id': 'req-node' }, body: 'body' });
    const second = await fetch(`${base}/demo?x=2`);
    expect(await first.text()).toBe('a');
    expect(await second.text()).toBe('b');
    expect(a.hits[0]).toContain('POST /demo?x=1 req-node');
    const health = await (await fetch(`${base}/__lb/health`)).json();
    const metrics = await (await fetch(`${base}/__lb/metrics`)).json();
    expect(health.backendSummary.healthy).toBe(2);
    expect(metrics.requestsTotal).toBe(2);
    await lb.shutdown();
  });

  it('supports backend pool add/remove and rejects invalid config', () => {
    const lb = new LoadBalancer(defaultConfig([{ id: 'a', url: 'http://a' }]));
    lb.addBackend({ id: 'b', url: 'http://b', weight: 3 });
    expect(lb.removeBackend('a')).toBe(true);
    expect(lb.snapshots().map((b) => b.id)).toEqual(['b']);
    expect(() => new LoadBalancer(defaultConfig([]))).toThrow(/backend/);
    expect(() => lb.addBackend({ id: 'bad', url: 'ftp://bad' })).toThrow(/http/);
  });
});

File: curriculum/11_load_balancer/go-impl/loadbalancer_test.go
package loadbalancer

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
	"time"
)

func TestRoundRobinWeightedEligibility(t *testing.T) {
	lb, err := New(DefaultConfig([]BackendConfig{{ID: "a", URL: "http://a", Weight: 2}, {ID: "b", URL: "http://b", Weight: 1}}))
	if err != nil {
		t.Fatal(err)
	}
	lb.MarkHealthy("a")
	lb.MarkHealthy("b")
	ids := []string{}
	for i := 0; i < 3; i++ {
		b, ok := lb.SelectBackend()
		if !ok {
			t.Fatal("missing backend")
		}
		ids = append(ids, b.ID)
	}
	if strings.Join(ids, ",") != "a,a,b" {
		t.Fatalf("weighted round robin = %v", ids)
	}
}

func TestLeastConnectionsAndPoolManagement(t *testing.T) {
	lb, err := New(Config{RoutingAlgorithm: LeastConnections, Backends: []BackendConfig{{ID: "b", URL: "http://b"}, {ID: "a", URL: "http://a"}}})
	if err != nil {
		t.Fatal(err)
	}
	lb.MarkHealthy("a")
	lb.MarkHealthy("b")
	selected, ok := lb.SelectBackend()
	if !ok || selected.ID != "a" {
		t.Fatalf("least connections selected %#v ok=%v", selected, ok)
	}
	if err := lb.AddBackend(BackendConfig{ID: "c", URL: "http://c", Weight: 3}); err != nil {
		t.Fatal(err)
	}
	if !lb.RemoveBackend("b") {
		t.Fatal("expected remove")
	}
	if len(lb.Snapshots()) != 2 {
		t.Fatalf("snapshots=%#v", lb.Snapshots())
	}
	if _, err := New(DefaultConfig(nil)); err == nil {
		t.Fatal("expected empty pool error")
	}
}

func TestHealthChecksAndCircuitBreaker(t *testing.T) {
	backend := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(http.StatusInternalServerError) }))
	defer backend.Close()
	lb, err := New(Config{Backends: []BackendConfig{{ID: "bad", URL: backend.URL}}, FailureThreshold: 1, UnhealthyThreshold: 1, HealthCheckTimeout: time.Second})
	if err != nil {
		t.Fatal(err)
	}
	lb.CheckBackend(context.Background(), "bad")
	s := lb.Snapshots()[0]
	if s.Health != Unhealthy || s.CircuitState != Open {
		t.Fatalf("snapshot=%#v", s)
	}
}

func TestSuccessfulHealthCheckMarksHealthyAndClosesHalfOpenCircuit(t *testing.T) {
	backend := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(http.StatusOK) }))
	defer backend.Close()
	lb, err := New(Config{Backends: []BackendConfig{{ID: "ok", URL: backend.URL}}, HealthyThreshold: 1, HealthCheckTimeout: time.Second})
	if err != nil {
		t.Fatal(err)
	}
	lb.mu.Lock()
	lb.backends["ok"].CircuitState = HalfOpen
	lb.mu.Unlock()
	lb.CheckBackend(context.Background(), "ok")
	s := lb.Snapshots()[0]
	if s.Health != Healthy || s.CircuitState != Closed {
		t.Fatalf("snapshot=%#v", s)
	}
}

func TestReverseProxyForwardsRequestAndAdminMetrics(t *testing.T) {
	var mu sync.Mutex
	hits := []string{}
	backendA := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/health" {
			w.WriteHeader(http.StatusOK)
			return
		}
		mu.Lock()
		hits = append(hits, "a "+r.Method+" "+r.URL.RequestURI()+" "+r.Header.Get("x-request-id"))
		mu.Unlock()
		w.Header().Set("x-backend", "a")
		_, _ = w.Write([]byte("a"))
	}))
	defer backendA.Close()
	backendB := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/health" {
			w.WriteHeader(http.StatusOK)
			return
		}
		mu.Lock()
		hits = append(hits, "b "+r.Method+" "+r.URL.RequestURI()+" "+r.Header.Get("x-request-id"))
		mu.Unlock()
		_, _ = w.Write([]byte("b"))
	}))
	defer backendB.Close()
	lb, err := New(DefaultConfig([]BackendConfig{{ID: "a", URL: backendA.URL}, {ID: "b", URL: backendB.URL}}))
	if err != nil {
		t.Fatal(err)
	}
	lb.MarkHealthy("a")
	lb.MarkHealthy("b")
	server := httptest.NewServer(lb)
	defer server.Close()
	req, _ := http.NewRequest(http.MethodPost, server.URL+"/demo?x=1", strings.NewReader("body"))
	req.Header.Set("x-request-id", "req-go")
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatal(err)
	}
	_ = res.Body.Close()
	res, err = http.Get(server.URL + "/demo?x=2")
	if err != nil {
		t.Fatal(err)
	}
	_ = res.Body.Close()
	mu.Lock()
	gotHits := strings.Join(hits, "|")
	mu.Unlock()
	if !strings.Contains(gotHits, "a POST /demo?x=1 req-go") || !strings.Contains(gotHits, "b GET /demo?x=2") {
		t.Fatalf("hits=%s", gotHits)
	}
	admin, err := http.Get(server.URL + "/__lb/health")
	if err != nil {
		t.Fatal(err)
	}
	defer admin.Body.Close()
	var body map[string]any
	if err := json.NewDecoder(admin.Body).Decode(&body); err != nil {
		t.Fatal(err)
	}
	if body["status"] != "ok" {
		t.Fatalf("admin=%#v", body)
	}
	metrics := lb.Metrics()
	if metrics.RequestsTotal != 2 || metrics.BackendRequests["a"] != 1 || metrics.BackendRequests["b"] != 1 {
		t.Fatalf("metrics=%#v", metrics)
	}
}

func TestProxyFailureAdminVariantsAndBasePathJoining(t *testing.T) {
	backend := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/demo" {
			t.Fatalf("expected joined upstream path /api/demo, got %s", r.URL.Path)
		}
		_, _ = w.Write([]byte("joined"))
	}))
	defer backend.Close()
	lb, err := New(Config{Backends: []BackendConfig{{ID: "joined", URL: backend.URL + "/api"}, {ID: "down", URL: "http://127.0.0.1:1"}}, FailureThreshold: 1})
	if err != nil {
		t.Fatal(err)
	}
	lb.MarkHealthy("joined")
	lb.MarkHealthy("down")
	server := httptest.NewServer(lb)
	defer server.Close()
	res, err := http.Get(server.URL + "/demo")
	if err != nil {
		t.Fatal(err)
	}
	_ = res.Body.Close()
	if res.StatusCode != http.StatusOK {
		t.Fatalf("joined proxy status=%d", res.StatusCode)
	}
	res, err = http.Get(server.URL + "/fails")
	if err != nil {
		t.Fatal(err)
	}
	_ = res.Body.Close()
	if res.StatusCode != http.StatusBadGateway {
		t.Fatalf("failed proxy status=%d", res.StatusCode)
	}
	for _, path := range []string{"/__lb/backends", "/__lb/metrics", "/__lb/missing"} {
		res, err := http.Get(server.URL + path)
		if err != nil {
			t.Fatal(err)
		}
		_ = res.Body.Close()
	}
	if lb.Snapshots()[1].CircuitState != Open {
		t.Fatalf("passive failure should open down circuit: %#v", lb.Snapshots()[1])
	}
}

func TestShutdownStopsHealthLoop(t *testing.T) {
	lb, err := New(DefaultConfig([]BackendConfig{{ID: "a", URL: "http://127.0.0.1:1"}}))
	if err != nil {
		t.Fatal(err)
	}
	lb.StartHealthChecks()
	if err := lb.Shutdown(context.Background()); err != nil {
		t.Fatal(err)
	}
}
